### ISSUE TITLE
Incorrect span order with Cloud Foundry

### ISSUE THREAD

**Issue by fbeaufume** [openzipkin/zipkin#2805]
**TLDR** : the traces from a microservices app (based on Spring Boot/Cloud/Sleuth/etc) are correctly displayed when the app is executed on a dev laptop (or Kubernetes), but have an incorrect order when executed on Cloud Foundry.

The application services : gateway (based on Spring Cloud Gateway), booking, flight, account  (the business services)

I extracted two JSON traces : **"success"** (when executed locally, shows the expected result) and **"failure"** (on CF, shows the problem), see attachements.

[zipkin_turbine_failure.json.txt](https://github.com/openzipkin/zipkin/files/3622459/zipkin_turbine_failure.json.txt)

[zipkin_turbine_success.json.txt](https://github.com/openzipkin/zipkin/files/3622460/zipkin_turbine_success.json.txt)

![zipkin_turbine_failure](https://user-images.githubusercontent.com/25341695/65060434-21548180-d978-11e9-8902-ba5ca032c60b.png)

![zipkin_turbine_success](https://user-images.githubusercontent.com/25341695/65060452-27e2f900-d978-11e9-9526-dfc5cbceec3e.png)

I spotted one difference : CF adds some "X-B3-TraceId=408ee7f252391ae6 X-B3-SpanId=408ee7f252391ae6" headers to the very first request ("gateway" in my case). But this alone does not seems to explain the issue, because if I force the same two B3 headers on my dev laptop, I still get correct spans ordering.

I can also provide some logs showing the traceId/spanId/parentSpanId on the various application services.

How can I get correct spans ordering, even when using Cloud Foundry ?

**Comment by codefromthecrypt** [comment]
This is troubleshooting in nature, and we've asked in our issue template to to gitter for that. I'll answer on gitter

**Comment by codefromthecrypt** [comment]
Here's the answer from gitter. I'm closing this issue. please do not reply-back as it spams watchers. The solution won't affect code in this repo so not nice to spam meanwhile.

--
for some reason your trace starts with a client span
https://github.com/openzipkin/zipkin/files/3622459/zipkin_turbine_failure.json.txt
ex "id": "408ee7f252391ae6", is the root and a CLIENT which I suppose is an odd decision in the gateway, but anyway, the SERVER side is missing the "shared" flag, which is present when a SERVER span re-uses the same ID as a client. This is an instrumentation bug. With that fixed, the order should be consistent. Check you are running all the orders, then check what is generating the SERVER span with "id": "408ee7f252391ae6",. This is not setting span.shared, which it should and something that brave's default middleware does. Once you identify that component, if it is sleuth and running the latest version of sleuth than raise a bug if still persists.

**Comment by codefromthecrypt** [comment]
A hint for where you go next, if the assumption is that the "post" span should have been the root, you can see it has its parent ID set.

This is why it is not showing up as root, because there's another root span ( the one I mentioned  the above comment.. the lack of shared flag isn't related eventhough it is a bug).

good luck

```
  {
    "traceId": "408ee7f252391ae6",
    "parentId": "408ee7f252391ae6",
    "id": "012080bf9d3bb2d1",
    "kind": "CLIENT",
    "name": "post",
    "timestamp": 1568723687373281,
    "duration": 67647,
    "localEndpoint": {
      "serviceName": "turbine-gateway",
      "ipv4": "x.y.237.146"
    },
    "tags": {
      "error": "409",
      "http.method": "POST",
      "http.path": "/bookings",
      "http.status_code": "409"
    }
  },
```
